pkg/scheduler/queue/queue_test.go

```diff
@@ -63,10 +78,149 @@ func makeSchedulerRequest(tenantID string) *SchedulerRequest {
+			startProducersChan := make(chan struct{})
+			producersErrGroup, _ := errgroup.WithContext(ctx)
+
+			runProducer := runQueueProducerIters(
+				queue, maxQueriersPerTenant, totalRequests/2, numProducers, numTenants, startProducersChan, queueDimensionFunc,
+			)
+			for producerIdx := 0; producerIdx < numProducers; producerIdx++ {
+				producerIdx := producerIdx
+				producersErrGroup.Go(func() error {
+					return runProducer(producerIdx)
+				})
+			}
+			close(startProducersChan)
+			err := producersErrGroup.Wait()
+			if err != nil {
+				require.NoError(t, err)
+			}
+		}
+
+		// emulate delay when consuming the slow queries
+		consumeFunc := func(request Request) error {
+			schedulerRequest := request.(*SchedulerRequest)
+			if schedulerRequest.AdditionalQueueDimensions[0] == slowConsumerQueueDimension {
+				time.Sleep(slowConsumerLatency)
+			}
+
+			queueTime := time.Since(schedulerRequest.EnqueueTime)
+			additionalQueueDimensionLabels := strings.Join(schedulerRequest.AdditionalQueueDimensions, ":")
+			queueDuration.With(prometheus.Labels{"additional_queue_dimensions": additionalQueueDimensionLabels}).Add(queueTime.Seconds())
+			return nil
+		}
+
+		// consume queries
+		queueConsumerErrGroup, ctx := errgroup.WithContext(ctx)
+		startConsumersChan := make(chan struct{})
+		runConsumer := runQueueConsumerIters(ctx, queue, totalRequests, numConsumers, startConsumersChan, consumeFunc)
+
+		for consumerIdx := 0; consumerIdx < numConsumers; consumerIdx++ {
+			consumerIdx := consumerIdx
+			queueConsumerErrGroup.Go(func() error {
+				return runConsumer(consumerIdx)
+			})
+		}
+
+		close(startConsumersChan)
+		err := queueConsumerErrGroup.Wait()
+		if err != nil {
+			require.NoError(t, err)
+		}
+
+		// record total queue duration by queue dimensions and whether the queue splitting was enabled
+		for _, queueDimension := range []string{normalQueueDimension, slowConsumerQueueDimension} {
+			queueDurationTotals[additionalQueueDimensionsEnabled][queueDimension] = promtest.ToFloat64(
+				queueDuration.With(prometheus.Labels{"additional_queue_dimensions": queueDimension}),
+			)
+		}
+
+		promRegistry.Unregister(queueDuration)
+	}
+
+	// total or average time in queue for a normal queue item should be roughly cut in half
+	// when queue splitting is enabled, as the average normal queue item waits behind
+	// half of the slow queue items, instead of waiting behind all the slow queue items.
+	expected := queueDurationTotals[false][normalQueueDimension] / 2
+	actual := queueDurationTotals[true][normalQueueDimension]
+	// some variance allowed due to actual time processing needed beyond the slow consumer delay;
+	// variance is also a function of the number of consumers and the consumer delay chosen.
+	// variance can be tighter if the test runs longer but there is a tradeoff for testing and CI speed
+	delta := expected * 0.10
+	require.InDelta(t, expected, actual, delta)
```

Comment:
A test that depends on timing makes me nervous... BUT I haven't been able to get it to flake locally in a few hundred runs :+1:

Reply:
there should be plenty of padding in the parameters here, essentially the delay time, number of consumers, and the allowed delta.
Changing only the number of consumers for example, don't see it start flaking until I push it up to 5 consumers. I've run 10,000 iterations with 4 consumers with no flakes. The code here only uses 1 consumer which has much more room for error.